### ISSUE TITLE
[Performance] Very long Typescript compile time

### ISSUE THREAD

**Issue by moufmouf** [stephenh/ts-proto#1135]
Hey!

When migrating from Typescript 5.0 to Typescript 5.1, I noticed a huge performance impact that happens specifically while compiling the files generated by ts-proto.

(the file used to take 6 seconds and suddenly, type checking is taking 110 seconds, which is a complete no-go for us).

I started by opening an issue on the Typescript repository. They closed it (because Typescript 5.1 is apparently doing the right thing and it's normal that it is taking more time).

See: https://github.com/microsoft/TypeScript/issues/60353

More specifically, the Typescript developers gave a look at the generated file and gave a few advices on how to speed up things (namely: replacing multiple "if" with "switch", which would both speed up Typescript compilation and runtime execution).

See: https://github.com/microsoft/TypeScript/issues/60353#issuecomment-2465522872

I thought it might be good to open this issue to show you this feedback.

Also, since the files are auto-generated by ts-proto, doing any type checking on those files is actually useless.

Is there any reason why you don't disable type checking at all by adding a ts-nocheck?

```typescript
//@ts-nocheck
```

@stephenh  I've seen you remove it here https://github.com/stephenh/ts-proto/pull/28#issuecomment-586752015 but I don't understand the rationale.

Note: Some other libraries (namely Relay, the GraphQL library) are putting @ts-nocheck by default on generated files and are considering adding a compiler flag to remove it (mostly for internally checking generated files are correctly typed: https://github.com/facebook/relay/pull/4753)

Final note: thanks a lot for ts-proto, it is so much better than the default protobuf JS generation. You guys rock! :heart:

**Comment by stephenh** [comment]
Hi @moufmouf ! Thanks for linking to the typescript issue, and also relay adding the `@ts-nocheck`...

I had removed `@ts-nocheck` back in the day, just because ts-proto has so many options :sweat_smile: that users can enable/combine in myriad of ways, I think its a good idea (by default, for most users) to have tsc go ahead and type-check the output.

That said, I'm not against changes to our output that would make tsc go faster...

I.e. reading Ryan's comment on `if (message.message?.$case === "batchMessage")` "should be using `switch`", I'm very open to that change--even more so if you'd like to submit a PR! :-)

These sort of "use `switch` instead of `if`" seems like a great/do-it-for-everyone win.

Maybe we start there, and see how the `tsc` compile times look with a `switch`?

After that, happy to look at adding `ts-nocheck` in certain places/if the user opts in to it with a `nocheck=true` type option...

**Comment by stephenh** [comment]
@moufmouf did you see this approach of putting `//@ts-nocheck` into the `*.proto` file as a comment, and that (At the time the comment was written anyway) would get output into the `*.ts` file:

https://github.com/stephenh/ts-proto/pull/28#issuecomment-586788525

Does that still work?

**Comment by moufmouf** [comment]
I gave this another shot and managed to implement the "switch case" solution instead of the "else if" solution.

It solves the performance issue!

Opening a PR right now!

**Comment by stephenh** [comment]
Ah gotcha.

Fwiw I opened up the example `*.proto` file you linked to, and I'm really surprised by how small it is. I.e. I assumed you were feeding in some enterprise ~100k-line long Frankenstein of a file, but really it seems pretty reasonable:

https://github.com/workadventure/workadventure/blob/develop/messages/protos/messages.proto

Which makes the "40 seconds for tsc" even more bizarre.

You do use `oneof`s a lot, but :thinking: still...

(I just saw your latest comment! Amazing!)
